### ISSUE TITLE
Thanks!

### ISSUE THREAD

**Issue by danielhanchen** [shivacharan22/Unsloth_Puzzles_sol#1]
Great work! I'll take a review of your submission more closely in the week (sorry been inundated with stuff!)

In general very good work - on FSDP - the training losses don't look correct - did you confirm if they are equivalent on single GPU training?

**Comment by shivacharan22** [comment]
Thanks for the time, Daniel—when I first noticed the loss discrepancies, I had the same concern, as I believe I tested on Kaggle. But I still re-ran it now on Kaggle with the P100 single GPU, and the losses seem to be similar(screenshot below).  The same code in colab with one T4 GPU gives different losses (screenshot below).

Kaggle P100 single_GPU losses:
<img width="884" alt="Image" src="https://github.com/user-attachments/assets/e9fa823a-a64b-43e7-bb8d-6feb78e5b3bf" />

colab T4 single_GPU losses: 
<img width="1272" alt="Image" src="https://github.com/user-attachments/assets/827c290e-c994-4d5b-94d9-97d05c9cf272" />

---------------------------------------------------------------------------------------------------
Also Note: **For the first question**,  if you are testing with an A100, please use my second version of the kernel (which I have commented), and set the block sizes to either 1024 or 2048. This gives me (1.46x) speedup, the first version of the kernel is also close to this speedup(1.32x) [both without torch.compile]. 
(I should have put the triton.autotune but just saw your tweet on A100 testing :))

<img width="720" alt="Image" src="https://github.com/user-attachments/assets/668bea8e-d845-4979-8234-dbc941280cd1" />

**Comment by shivacharan22** [comment]
Hello @danielhanchen , I hope you're having a great day. I was wondering if you've had a chance to review my solutions. I’d really appreciate any feedback you have, especially regarding how I scored. Thanks so much for your time!

**Comment by danielhanchen** [comment]
Apologies on the delay!

1. Task A - I tried version 1 and 2 - 2 sadly is unfortunately much slower than normal dequantization (1 is sadly even slower!) - on a T4 - it's 14% slower, and L4 10% slower - great attempt though! Your 2nd version without using another buffer for the double dequant is the faster of the 2 - but both are still slow sorry! I think the use of inline asm might actually be confusing the compiler - so it's probably not ideal!
2. For Task 2 - oh I meant https://github.com/shivacharan22/Unsloth_Puzzles_sol/blob/main/unsloth-q2.ipynb - your training losses are abnormally high (9 to 10) is definitely worth more investigation!

Overall great attempt and great work - but sadly there are some issues with your submission - thanks again!

**Comment by shivacharan22** [comment]
Thanks for the feedback, Daniel—I appreciate it, especially since I know you have a lot to attend to.

Regarding Task 1, did you have a chance to check the performance on the A100? I'm seeing significant improvements there with V2.

For Task 2, I suspect the difference might be due to the batch size. I've reduced it to align with the typical loss values you observed. Aside from that, single-GPU training without my modifications produces similar results; you can verify this in version 2 on Kaggle or the screenshots provided above.

Also, have you had an opportunity to review Task 3 and Task 5? For Task 5, I've included the plot for your reference.

**Comment by danielhanchen** [comment]
For Task C - you actually showcase the losses (it it should be 1.5 not 9 to 10) It's not a batch size issue, since that actually should not explode the loss.
```
Step | Training Loss
-- | --
1 | 1.519300
2 | 2.393000
3 | 2.499000
4 | 3.531300
```
You did actually OK on task C, but your own Jupyter notebook showed you only compiled 2 functions - I'm assuming you might have reran your notebook?

Task 5 is cool! I can see you're using `torch.func.grad_and_value` which is smart! But again your training loss is wrong:
```
1 | 3.296300
2 | 5.318100
3 | 3.403900
4 | 6.565700
5 | 4.644800
```
In fact your own plot shows it's different - it's supposed to nearly mostly match:
![Image](https://github.com/user-attachments/assets/1276903a-09b5-42f8-8aed-a7f23a6762fd)

Appreciate the effort though - great work!
